examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,577 @@
+    loss=losses,
+)
+
+# Begin training the model.
+clustering_learner.fit(x=inputs, y=labels, batch_size=512, epochs=50)
+
+"""
+Plot training loss
+"""
+
+plt.plot(history.history["loss"])
+plt.ylabel("loss")
+plt.xlabel("epoch")
+plt.show()
+
+"""
+## Cluster analysis
+"""
+
+"""
+### Assign images to clusters
+"""
+
+# Get the cluster probability distribution of the input images.
+clustering_probs = clustering_model.predict(x_data, batch_size=batch_size, verbose=1)
+# Get the cluster of the highest probability.
+cluster_assignments = tf.math.argmax(clustering_probs, axis=-1).numpy()
+# Store the clustering confidence.
+# Images with the highest clustering confidence are considered the 'prototypes'
+# of the clusters.
+cluster_confidence = tf.math.reduce_max(clustering_probs, axis=-1).numpy()
+
+"""
+Let's compute the cluster sizes
+"""
+
+clusters = defaultdict(list)
+for idx, c in enumerate(cluster_assignments):
+    clusters[c].append((idx, cluster_confidence[idx]))
+
+for c in range(num_clusters):
+    print("cluster", c, ":", len(clusters[c]))
+
+"""
+Notice that the clusters have roughly balanced sizes.
+"""
+
+"""
+### Visualize cluster images
+
+Display the *prototypes*—instances with the highest clustering confidence—of each cluster:
+"""
+
+num_images = 8
+plt.figure(figsize=(15, 15))
+position = 1
+for c in range(num_clusters):
+    cluster_instances = sorted(clusters[c], key=lambda kv: kv[1], reverse=True)
+
+    for j in range(num_images):
+        image_idx = cluster_instances[j][0]
+        plt.subplot(num_clusters, num_images, position)
+        plt.imshow(x_data[image_idx].astype("uint8"))
+        plt.title(class_labels[y_data[image_idx][0]])
+        plt.axis("off")
+        position += 1
+
+"""
+### Compute clustering accuracy
+
+First, we assign a label for each cluster based on the majority label of its images.
+Then, we compute the accuracy of each cluster by dividing the number of image with the
+majority label by the size of the cluster.
+"""
+
+cluster_label_counts = dict()
+
+for c in range(num_clusters):
+    cluster_label_counts[c] = [0] * num_classes
+    instances = clusters[c]
+    for i, _ in instances:
+        cluster_label_counts[c][y_data[i][0]] += 1
+
+    cluster_label_idx = np.argmax(cluster_label_counts[c])
+    correct_count = np.max(cluster_label_counts[c])
+    cluster_size = len(clusters[c])
+    accuracy = (
+        np.round((correct_count / cluster_size) * 100, 2) if cluster_size > 0 else 0
+    )
+    cluster_label = class_labels[cluster_label_idx]
+    print("cluster", c, "label is:", cluster_label, " -  accuracy:", accuracy, "%")
+
+"""
+## Conclusion
+
+To improve the accuracy results, you can: 1) increase the number
+of epochs in the representation learning and the clustering phases, 2)
+allow the encoder weights to be tuned during the clustering phase, and 3) perform a final
+fine-tuning step through self-labeling, as described in the [paper](https://arxiv.org/abs/2005.12320).
```

Comment:
Suggestion:

- Use semi-colons instead of commas here


```suggestion
To improve the accuracy results, you can: 1) increase the number
of epochs in the representation learning and the clustering phases; 2)
allow the encoder weights to be tuned during the clustering phase; and 3) perform a final
fine-tuning step through self-labeling, as described in the [original SCAN paper](https://arxiv.org/abs/2005.12320).
```